src/librustc_mir/hair/cx/expr.rs

```diff
@@ -76,12 +96,7 @@ fn apply_adjustment<'a, 'gcx, 'tcx>(cx: &mut Cx<'a, 'gcx, 'tcx>,
     let Expr { temp_lifetime, mut span, .. } = expr;
     let kind = match adjustment.kind {
         Adjust::Pointer(PointerCast::Unsize) => {
-            if let ExprKind::Block { body } = expr.kind {
-                if let Some(ref last_expr) = body.expr {
-                    span = last_expr.span;
-                    expr.span = span;
-                }
-            }
+            span = adjust_span(&mut expr);
```

Comment:
Does this actually compile? I'd expect a borrowck error here

Ok, I am beyond confused by this code now. `span` is essentially `&mut expr.span`. So `expr.span = span` is a nop I guess? Looks to me like `expr.span = span` can be removed entirely, so your `adjust_span` doesn't need to return a `Span` anymore.

Reply:
> Does this actually compile? I'd expect a borrowck error here

Tried a similar example
```rust
#[derive(Debug, Clone)]
struct X {
    aa: String
}

fn main() {
   let mut x = X{aa: "asd".into()};
   
   let X {aa} = x;
   
   test(&mut x);
}

fn test(x: &mut X) {
    x.aa = "def".into();
}
```

This is the error
```
error[E0382]: borrow of moved value: `x`
  --> src/main.rs:11:9
   |
9  |    let X {aa} = x;
   |           -- value moved here
10 |    
11 |    test(&mut x);
   |         ^^^^^^ value borrowed here after partial move
   |
   = note: move occurs because `x.aa` has type `std::string::String`, which does not implement the `Copy` trait
```

I'm not able to understand why the code I have written compiles :/

Reply:
Oh got it! It compiles because `Span` is a `u32`.

Reply:
Got it! The code compiles because in https://github.com/rust-lang/rust/blob/999ee0190fbcabf13253d6f95bb7ab73709c9240/src/librustc_mir/hair/cx/expr.rs#L96 `temp_lifetime` and `span` are `Copy`.

Reply:
ah of course. Maybe make the helper function a closure, so it has access to `span`?

Reply:
Done.